src/analysis/normed_space/deriv.lean

```diff
@@ -0,0 +1,178 @@
+/-
+Copyright (c) 2019 Jeremy Avigad. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Jeremy Avigad
+
+The Fréchet derivative.
+
+Let `E` and `F` be normed spaces, and `f : E → F`. Then
+
+  `has_fderiv_at_within f f' x s`
+
+says that the function `f' : E → F` is a bounded linear map and `f` has derivative `f'` at
+`x`, where the domain of interest is restricted to `s`. We also have
+
+  `has_fderiv_at f f' x := has_fderiv_at_within f f' x univ`
+
+We may want to prove some theorems specifically for this special case, but most of the more
+general theorems should apply right out of the box.
+
+The derivative is characterized in two ways: in terms of the `tendsto` relation, and in terms of
+the `littleo` relation.
+-/
+import topology.basic analysis.normed_space.bounded_linear_maps ..asymptotics tactic.abel
+open filter asymptotics
+
+variables {E : Type*} [normed_space ℝ E]
+variables {F : Type*} [normed_space ℝ F]
+variables {G : Type*} [normed_space ℝ G]
+
+def has_fderiv_at_within (f : E → F) (f' : E → F) (x : E) (s : set E) :=
+is_bounded_linear_map f' ∧
+  tendsto (λ x', ∥x' - x∥⁻¹ • (f x' - f x - f' (x' - x))) (nhds_within x s) (nhds 0)
+
+def has_fderiv_at (f : E → F) (f' : E → F) (x : E) :=
+has_fderiv_at_within f f' x set.univ
```

Comment:
This is the definition I was talking about yesterday. You should define

```
has_deriv_at f f' x :=
is_bounded_linear_map f' ∧
  tendsto (λ x', ∥x' - x∥⁻¹ • (f x' - f x - f' (x' - x))) (nhds x) (nhds 0)
```
and then prove this equality.